src/NumericalAlgorithms/Interpolation/Callbacks/FindApparentHorizon.hpp

```diff
@@ -0,0 +1,200 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <utility>
+
+#include "ApparentHorizons/FastFlow.hpp"
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "Informer/Verbosity.hpp"
+#include "Parallel/Abort.hpp"
+#include "Parallel/ConstGlobalCache.hpp"
+#include "Parallel/Invoke.hpp"
+#include "Parallel/Printf.hpp"
+#include "PointwiseFunctions/GeneralRelativity/Tags.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/PrettyType.hpp"
+
+/// \cond
+namespace StrahlkorperTags {
+template <typename Frame>
+struct Strahlkorper;
+}  // namespace StrahlkorperTags
+namespace db {
+template <typename TagsList>
+class DataBox;
+}  // namespace db
+namespace intrp {
+template <class Metavariables, typename InterpolationTargetTag>
+struct InterpolationTarget;
+namespace Tags {
+template <typename Metavariables>
+struct TemporalIds;
+}  // namespace Tags
+}  // namespace intrp
+namespace ah {
+namespace Tags {
+struct FastFlow;
+}  // namespace Tags
+}  // namespace ah
+namespace Tags {
+struct Verbosity;
+}  // namespace Tags
+template <typename Frame>
+class Strahlkorper;
+/// \endcond
+
+namespace intrp {
+namespace callbacks {
+
+/// \brief post interpolation callback (see InterpolationTarget)
+/// that does a FastFlow iteration and triggers another one until convergence.
+///
+/// Assumes that InterpolationTargetTag contains an additional
+/// struct called `post_horizon_find_callback`, which has a function
+///```
+///  static void apply(const DataBox<DbTags>&,
+///                    const intrp::ConstGlobalCache<Metavariables>&,
+///                    const Metavariables::temporal_id&) noexcept;
+///```
+/// that is called if the FastFlow iteration has converged.
+///
+/// Uses:
+/// - Metavariables:
+///   - `temporal_id`
+///   - `domain_frame`
+/// - DataBox:
+///   - `::Tags::Verbosity`
+///   - `::gr::Tags::InverseSpatialMetric<3,Frame>`
+///   - `::gr::Tags::ExtrinsicCurvature<3,Frame>`
+///   - `::gr::Tags::SpatialChristoffelSecondKind<3,Frame>`
+///   - `::ah::Tags::FastFlow`
+///   - `StrahlkorperTags::Strahlkorper<Frame>`
+///
+/// Modifies:
+/// - DataBox:
+///   - `::ah::Tags::FastFlow`
+///   - `StrahlkorperTags::Strahlkorper<Frame>`
+///
+/// This is an InterpolationTargetTag::post_interpolation_callback;
+/// see InterpolationTarget for a description of InterpolationTargetTag.
+template <typename InterpolationTargetTag>
+struct FindApparentHorizon {
```

Comment:
This action would be particularly important to test individually. With the mocking framework it should decouple nicely from everything else.

Reply:
I'm not sure what is being asked here.  Test_ApparentHorizonFinder **is** a test of this callback,
iterated until convergence. 

This entire callback basically does 
`if(converged) {A(); return true;} else {B(); return false;}`
When computing `converged`, it also changes the shape of the trial surface (inside the call to `fast_flow->iterate_horizon_finder`), but there is no meaningful way to compute what the actual trial surface *should* be after an iteration, other than calling `fast_flow->iterate_horizon_finder` in the test, which doesn't test anything because I'm just calling the same function twice.

So the only thing I could meaningfully do is test the `if` statement. One way to do this is set up a mocking structure to call `FindApparentHorizon`, and pass it something that I know is very far from an apparent horizon (e.g. give it a spherical initial guess on a spin 0.9 Kerr hole), and then test that some mock version of `B()` is called and that it returns false.  Or pass it the actual solution (sphere of radius 2 for Schwarzschild), and test that some mock version of `A()` is called and that it returns true.  Should I do this?

Reply:
Also note that `FindApparentHorizon` is not an Action. It's a callback.  So it doesn't really decouple. So testing the `if` statement independently as described above would be mostly a copy of `Test_ApparentHorizonFinder` but run only one iteration and checking only whether or not it converged in that iteration.

Reply:
Decided during the Wednesday call, not to add this test

Reply:
ok